src/main/c/Objects/pyjmap.c

```diff
@@ -223,6 +218,153 @@ static PyObject* pyjmap_getiter(PyObject* obj)
     return result;
 }
 
+static PyObject* pyjmap_keys(PyObject* self, PyObject* args)
+{
+    jobject    keyset = NULL;
+    PyObject  *result = NULL;
+    PyJObject *pyjob  = (PyJObject*) self;
+    JNIEnv    *env    = pyembed_get_env();
+
+    if ((*env)->PushLocalFrame(env, JLOCAL_REFS) != 0) {
+        process_java_exception(env);
+        return NULL;
+    }
+
+    keyset = java_util_Map_keySet(env, pyjob->object);
+    if (process_java_exception(env) || !keyset) {
+        goto FINALLY;
+    }
+
+    result = jobject_As_PyObject(env, keyset);
+FINALLY:
+    (*env)->PopLocalFrame(env, NULL);
+    return result;
+}
+
+static PyObject* pyjmap_items(PyObject* self, PyObject* args)
+{
+    jobject    entrySet = NULL;
+    jobject    itr      = NULL;
+    PyObject  *pylist   = NULL;
+    PyObject  *result   = NULL;
+    int        size     = 0;
+    int        index    = 0;
+    PyJObject *pyjob    = (PyJObject*) self;
+    JNIEnv    *env      = pyembed_get_env();
+
+    if ((*env)->PushLocalFrame(env, JLOCAL_REFS) != 0) {
+        process_java_exception(env);
+    }
+
+    entrySet = java_util_Map_entrySet(env, pyjob->object);
+    if (process_java_exception(env) || !entrySet) {
+        goto FINALLY;
+    }
+
+    size = java_util_Map_size(env, pyjob->object);
+    if (process_java_exception(env)) {
+        goto FINALLY;
+    }
+
+    itr = java_lang_Iterable_iterator(env, entrySet);
+    if (process_java_exception(env) || !itr) {
+        goto FINALLY;
+    }
+
+    pylist = PyList_New(size);
+    while (java_util_Iterator_hasNext(env, itr)) {
```

Comment:
I suspect it would be better to do a for loop over size here. The reason is because the documentation on PyList_New is very clear that you have to call setItem for every item so if the java iterator didn't return size items it would get crashy. I don't think there is any normal case where this would happen but I get a little nervous about concurrent modification or poorly written maps.

Reply:
Done